File: Algorithm/Two Pointers/876. Middle of the Linked List /MiddleOfTheLinkedList.cpp
//
//  MiddleOfTheLinkedList.cpp
//  LeetCode-main
//
//  Created by jackma on 2022/4/8.
//  Copyright © 2022 JackMa. All rights reserved.
//

#include "MiddleOfTheLinkedList.hpp"

/**
 Given the head of a singly linked list, return the middle node of the linked list.

 If there are two middle nodes, return the second middle node.
 
 Input: head = [1,2,3,4,5]
 Output: [3,4,5]
 Explanation: The middle node of the list is node 3.
 
 Input: head = [1,2,3,4,5,6]
 Output: [4,5,6]
 Explanation: Since the list has two middle nodes with values 3 and 4, we return the second one.
 */

void testMiddleOfTheLinkedList() {
    vector<int> nums = {1};
    outputList(Solution().middleNode(buildList(nums)));
}

ListNode* Solution::middleNode(ListNode *head) {
    ListNode *newHead = new ListNode();
    newHead->next = head;
    // two pointer
    ListNode *fast = newHead, *slow = newHead;
    
    while (fast != NULL) {
        slow = slow->next;
        fast = fast->next ? fast->next->next : NULL;
    }
    
    return slow;
}

File: Algorithm/Two Pointers/876. Middle of the Linked List /MiddleOfTheLinkedList .swift
//
//  MiddleOfTheLinkedList .swift
//  LeetCode-Swift
//
//  Created by jackma on 2023/3/9.
//  Copyright © 2023 JackMa. All rights reserved.
//

import Foundation

/**
 https://leetcode.com/problems/middle-of-the-linked-list
 
 Given the head of a singly linked list, return the middle node of the linked list.

 If there are two middle nodes, return the second middle node.
 */

extension Solution {
    func MiddleOfTheLinkedList(_ head: ListNode?) -> ListNode? {
        if head == nil {
            return head
        }
        let zeroHead = ListNode(0, head)
        var curr:Optional = zeroHead, prev:Optional = zeroHead
        while curr != nil || prev != nil {
            // prev is the middle of list
            if curr == nil {
                return prev
            }
            // curr speed is 2x
            if let currNext = curr!.next {
                curr = currNext.next
            } else {
                curr = nil
            }
            // prev speed is 1x
            if let prevNext = prev!.next {
                prev = prevNext
            }
        }
        
        return nil
    }
}
